src/System.Buffers.Experimental/System/Buffers/ReadWriteBytes.cs

```diff
@@ -0,0 +1,322 @@
+                {
+                    return Position.Create(first.Length - index);
+                }
+                else
+                {
+                    var allIndex = index + (_all.Memory.Length - first.Length);
+                    return Position.Create(allIndex, _all);
+                }
+            }
+            if (Rest == null) return default;
+            return PositionOf(Rest, value);
+        }
+
+        public Position PositionAt(int index)
+        {
+            ReadOnlySpan<byte> first = _first.Span;
+            int firstLength = first.Length;
+
+            if (index < firstLength)
+            {
+                if (_all == null)
+                {
+                    return Position.Create(firstLength - index);
+                }
+                else
+                {
+                    var allIndex = index + (_all.Memory.Length - firstLength);
+                    return Position.Create(allIndex, _all);
+                }
+            }
+            if (Rest == null) return default;
+            return PositionAt(Rest, index - firstLength);
+        }
+
+        private static Position PositionOf(IMemoryList<byte> list, byte value)
+        {
+            ReadOnlySpan<byte> first = list.Memory.Span;
+            int index = first.IndexOf(value);
+            if (index != -1) return Position.Create(index, list);
+            if (list.Rest == null) return default;
+            return PositionOf(list.Rest, value);
+        }
+
+        private static Position PositionAt(IMemoryList<byte> list, int index)
+        {
+            if (list == null) return default;
+            ReadOnlySpan<byte> first = list.Memory.Span;
+            int firstLength = first.Length;
+
+            if (index < firstLength)
+            {
+                return Position.Create(index, list);
+            }
+            return PositionAt(list.Rest, index - firstLength);
+        }
+
+        public int CopyTo(Span<byte> buffer)
+        {
+            var first = First;
+            var firstLength = first.Length;
+            if (firstLength > buffer.Length)
+            {
+                first.Slice(0, buffer.Length).Span.CopyTo(buffer);
+                return buffer.Length;
+            }
+            first.Span.CopyTo(buffer);
+            if (buffer.Length == firstLength || Rest == null) return firstLength;
+            return firstLength + Rest.CopyTo(buffer.Slice(firstLength));
+        }
+
+        ReadWriteBytes SliceRest(long index, long length)
+        {
+            if (Rest == null)
+            {
+                if (First.Length == index && length == 0)
+                {
+                    return Empty;
+                }
+                else
+                {
+                    throw new ArgumentOutOfRangeException("index or length");
+                }
+            }
+
+            // TODO (pri 2): this could be optimized
+            var rest = new ReadWriteBytes(Rest, length + index - _first.Length);
```

Comment:
Do we need to validate the index/length here?

Reply:
I added validation to Slice(int,int)